src/asl.ts

```diff
@@ -2837,60 +2792,494 @@ export class ASL {
               : undefined;
 
             if (!propName) {
-              throw new SynthError(
-                ErrorCodes.StepFunctions_property_names_must_be_constant
-              );
+              // step function does not support variable property accessors
+              // this will probably fail in the filter to ASLGraph implementation too
+              // however, lets let ASLGraph try and fail if needed.
+              return undefined;
             }
 
-            return `${resolveRef(ref.parent.parent)}[${propName}]`;
-          } else if (
-            isVariableDecl(ref) ||
-            isFunctionDecl(ref) ||
-            isSetAccessorDecl(ref) ||
-            isGetAccessorDecl(ref) ||
-            isClassDecl(ref) ||
-            isClassMember(ref)
-          ) {
-            throw new SynthError(
-              ErrorCodes.StepFunction_invalid_filter_syntax,
-              `cannot reference a ${ref.kindName} within a JSONPath .filter expression`
-            );
+            return `${value}['${propName}']`;
           }
-          assertNever(ref);
         }
       } else if (isPropAccessExpr(expr)) {
-        return `${toFilterCondition(expr.expr)}.${expr.name.name}`;
+        const value = toFilterCondition(expr.expr);
+        return value ? `${value}.${expr.name.name}` : undefined;
       } else if (isElementAccessExpr(expr)) {
-        return `${toFilterCondition(
-          expr.expr
-        )}[${this.assertElementAccessConstant(
+        const field = this.assertElementAccessConstant(
           ASLGraph.getAslStateOutput(this.eval(expr.element))
-        )}]`;
+        );
+
+        const value = toFilterCondition(expr.expr);
+
+        return value
+          ? `${value}[${typeof field === "number" ? field : `'${field}'`}]`
+          : undefined;
       }
 
+      // try to compile with ASLGraph instead
+      return undefined;
+    };
+
+    const expression = toFilterCondition(stmt.expr);
+    return expression
+      ? {
+          jsonPath: `${valueJsonPath.jsonPath}[?(${expression})]`,
+        }
+      : undefined;
+  }
+
+  /**
+   * filter an array using ASLGraph. Should support any logic that returns a boolean already supported by Functionless Step Functions.
+   *
+   * Note: If possible, try to use {@link filterToJsonPath} as it introduces no new state transitions.
+   */
+  private filterToASLGraph(
+    valueJsonPath: ASLGraph.JsonPath,
+    predicate: FunctionExpr | ArrowFunctionExpr
+  ): ASLGraph.OutputSubState {
+    const predicateResult = this.newHeapVariable();
+    const filterResult = this.newHeapVariable();
+
+    const [itemParameter, indexParameter, arrayParameter] =
+      predicate.parameters;
+
+    const assignParameters = ASLGraph.joinSubStates(
+      predicate,
+      itemParameter
+        ? this.evalDecl(itemParameter, {
+            jsonPath: indexParameter
+              ? `${filterResult}.arr[0].item`
+              : `${filterResult}.arr[0]`,
```

Comment:
It's not clear to me why we are appending these to the json paths.

Reply:
Will document. 

Summary: in order to get the index, we zip the array items with their index. `[{ index: 0, item: item1], [index: 1, item : item2}, ...]`

When we zip, we need to reference `[].item` or `[].index` instead of just the array position. 

But we only do this when needed.